flow/display_list.cc

```diff
@@ -0,0 +1,1248 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <type_traits>
+
+#include "flutter/flow/display_list.h"
+#include "flutter/flow/display_list_canvas.h"
+#include "flutter/flow/display_list_utils.h"
+#include "flutter/fml/logging.h"
+
+#include "third_party/skia/include/core/SkImageFilter.h"
+#include "third_party/skia/include/core/SkMaskFilter.h"
+#include "third_party/skia/include/core/SkPath.h"
+#include "third_party/skia/include/core/SkRSXform.h"
+#include "third_party/skia/include/core/SkTextBlob.h"
+
+// This header file cannot be included here, but we cannot
+// record calls made by the SkShadowUtils without it.
+// #include "third_party/skia/src/core/SkDrawShadowInfo.h"
+
+namespace flutter {
+
+const SkSamplingOptions DisplayList::NearestSampling =
+    SkSamplingOptions(SkFilterMode::kNearest, SkMipmapMode::kNone);
+const SkSamplingOptions DisplayList::LinearSampling =
+    SkSamplingOptions(SkFilterMode::kLinear, SkMipmapMode::kNone);
+const SkSamplingOptions DisplayList::MipmapSampling =
+    SkSamplingOptions(SkFilterMode::kLinear, SkMipmapMode::kLinear);
+const SkSamplingOptions DisplayList::CubicSampling =
+    SkSamplingOptions(SkCubicResampler{1 / 3.0f, 1 / 3.0f});
+
+enum dlCompare {
+  NOT_EQUAL,
+  BULK_COMPARE,
```

Comment:
Can you document what bulk compare means?

Reply:
I added this:
```
  // The Op is deferring comparisons to a bulk memcmp performed lazily
  // across all bulk-comparable ops. Most ops fall into this category
  // if they contain raw numeric values or have sk_sp<> pointers that
  // point to objects that we cannot compare explicitly and so we rely
  // on their identities to compare ==. (Note that two sk_sp<> that
  // point to the same object are themselves identical ==.)
  kUseBulkCompare,
```

Reply:
I reworked it a bit more after that with a comment on the reason for the enum to exist in the first place. Did it explain things well enough?